File: functions/api/api-session.js
function ApiSession(api, database, model)
{
	const shajs = require('sha.js')

	this.login = function(request, response)
	{
		const email    = request.body.email
		const password = request.body.password

		if (email && password)
		{
			database.account.byEmail(email)
			.then(accountDoc =>
			{
				if (accountDoc.hasPassword(hash(shajs, password)))
				{
					const seed = Math.random().toString(36).substring(2) + new Date().getTime().toString(36)
					const sessionId = hash(shajs, seed)

					accountDoc.update({session: sessionId})
					.then(result =>
					{
						response
							.status(200)
							.set(api.SESSION_TOKEN, sessionId)
							.send()
					})
				}
				else
				{
					response.status(401).send()
				}
			})
			.catch(error =>
			{
				response.status(404).send()
			})
		}
		else
		{
			response.status(400).send()
		}
	}

	function hash(shajs, input)
	{
		return new shajs.sha512().update(input).digest('hex')
	}
}

module.exports = ApiSession

File: src/main/kotlin/com/mauriciotogneri/momowars/api/ApiSession.kt
package com.mauriciotogneri.momowars.api

import com.mauriciotogneri.momowars.exception.UnauthorizedException
import com.mauriciotogneri.momowars.express.Request
import com.mauriciotogneri.momowars.express.Response
import com.mauriciotogneri.momowars.express.bodyParam
import com.mauriciotogneri.momowars.utils.Hash
import kotlin.js.Date
import kotlin.js.Math
import kotlin.js.json

class ApiSession : BaseApi()
{
    fun createSession(request: Request, response: Response)
    {
        process(response) { database ->

            val email = request.bodyParam("email")
            val password = request.bodyParam("password")

            checkNotEmpty(email, password)

            val documentAccount = database.account.byEmail(email)

            if (!documentAccount.hasPassword(Hash.sha512(password)))
            {
                throw UnauthorizedException()
            }

            val sessionId = newSessionId()

            val json = json()
            json["session"] = sessionId

            documentAccount.update(json)

            response
                    .status(CREATED)
                    .set(Api.SESSION_TOKEN, sessionId)
                    .send()
        }
    }

    fun newSessionId(): String
    {
        val seed = Math.random().toString().substring(2) + Date().getTime().toString()

        return Hash.sha512(seed)
    }
}
